pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,259 @@
+		DeleteFunc: func(obj interface{}) { c.enqueueCluster(obj) },
+	})
+	c.clusterLister = csif.Cluster().V1alpha1().Clusters().Lister()
+	csif.WaitForCacheSync(stopCh)
+	csif.Start(stopCh)
+
+	nsif := informers.NewSharedInformerFactoryWithOptions(kubeClient, resyncPeriod)
+	nsif.Core().V1().Namespaces().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueNamespace(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueueNamespace(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueNamespace(obj) },
+	})
+	c.namespaceLister = nsif.Core().V1().Namespaces().Lister()
+	nsif.WaitForCacheSync(stopCh)
+	nsif.Start(stopCh)
+
+	return c
+}
+
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	dynClient  dynamic.Interface
+	kubeClient kubernetes.Interface
+
+	clusterLister   clusterlisters.ClusterLister
+	namespaceLister corelister.NamespaceLister
+	namespaceClient coreclient.NamespaceInterface
+	stopCh          chan struct{}
+	indexer         cache.Indexer
+	lister          cache.GenericLister
+	gvkTrans        *gvk.GVKTranslator
+}
+
+func (c *Controller) enqueueResource(gvr schema.GroupVersionResource, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited(gvr.String() + "::" + key)
+}
+
+func (c *Controller) enqueueNamespace(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("namespace::" + key)
+}
+
+func (c *Controller) enqueueCluster(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("cluster::" + key)
+}
+
+func (c *Controller) Start(numThreads int) {
+	defer c.queue.ShutDown()
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(c.startWorker, time.Second, c.stopCh)
+	}
+	klog.Infof("Starting workers")
+	<-c.stopCh
+	klog.Infof("Stopping workers")
+}
+
+func (c *Controller) startWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	var err error
+	if strings.HasPrefix(key, "namespace::") {
+		err = c.processNamespace(strings.TrimPrefix(key, "namespace::"))
+	} else if strings.HasPrefix(key, "cluster::") {
+		err = c.processCluster(strings.TrimPrefix(key, "cluster::"))
+	} else {
+		err = c.process(key)
+	}
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
```

Comment:
I think I got this bit from Knative, which Tekton also reuses. My main reason for wanting to retry is if a lister cache doesn't have a namespace or cluster yet, we shouldn't fail after one try, but we also should also give up eventually.

If you have a better pattern for this I'm open to ideas, we should do the same in cluster controller and deployment splitter too.

Reply:
Won't you re-try on re-list?

Reply:
> If you have a better pattern for this I'm open to ideas, we should do the same in cluster controller and deployment splitter too.

+1 I've never been very comfortable with the idea of stopping after a fix number of retries. IIRC we already had to bypass this behavior with a specific `RetryableError` type in some places in the cluster controller to avoid loosing an event after a number of retries.

Reply:
I'm sold. I'll use RetryableError and not give up after 5 retries. I've also cleaned up RetryableError usage in the other packages.